Terraform crash on aws_emr_instance_group resize

_This issue was originally opened by @synhershko as hashicorp/terraform#15649. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Using Terraform v0.10.0-rc1-dev on linux mint 64bit, I executed terraform apply setting an aws_emr_instance_group from 5 to 0, and got the following crash:

```
aws_emr_instance_group.core-nodes: Still modifying... (ID: ig-2PW30KB5TCX7N, 10s elapsed)
Error applying plan:

1 error(s) occurred:

* aws_emr_instance_group.core-nodes: 1 error(s) occurred:

* aws_emr_instance_group.core-nodes: unexpected EOF

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
panic: interface conversion: interface {} is *emr.EMR, not *aws.AWSClient
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: 
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: goroutine 445 [running]:
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: github.com/terraform-providers/terraform-provider-aws/aws.fetchAllEMRInstanceGroups(0x1fbbde0, 0xc4203fc310, 0xc42041ea00, 0xf, 0x17dc1, 0x1625e7da, 0xc42004ec58, 0x4457bb, 0x32b5dc0)
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_emr_instance_group.go:190 +0x750
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: github.com/terraform-providers/terraform-provider-aws/aws.fetchEMRInstanceGroup(0x1fbbde0, 0xc4203fc310, 0xc42041ea00, 0xf, 0xc42041e9a0, 0x10, 0xc42004ece0, 0x139a824, 0x18)
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_emr_instance_group.go:225 +0x56
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: github.com/terraform-providers/terraform-provider-aws/aws.instanceGroupStateRefresh.func1(0xc42004ef20, 0xc42066af60, 0x0, 0x1, 0xc4204260a0, 0x2)
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_emr_instance_group.go:285 +0x5b
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource.(*StateChangeConf).WaitForState.func1(0xc420664960, 0xc4201b1ea0, 0xc4206a9200, 0xc4206649c0, 0xc42041ab00, 0xc42041aa78)
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource/state.go:103 +0x1c8
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: created by github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource.(*StateChangeConf).WaitForState
2017/07/27 21:01:06 [DEBUG] plugin: terraform-provider-aws_v0.1.2_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource/state.go:195 +0x249
2017/07/27 21:01:06 [DEBUG] root: eval: *terraform.EvalWriteState
2017/07/27 21:01:06 [DEBUG] root: eval: *terraform.EvalApplyProvisioners
2017/07/27 21:01:06 [DEBUG] root: eval: *terraform.EvalIf
2017/07/27 21:01:06 [DEBUG] root: eval: *terraform.EvalWriteState
2017/07/27 21:01:06 [DEBUG] root: eval: *terraform.EvalWriteDiff
2017/07/27 21:01:06 [DEBUG] root: eval: *terraform.EvalApplyPost
2017/07/27 21:01:06 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* aws_emr_instance_group.core-nodes: unexpected EOF
2017/07/27 21:01:06 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* aws_emr_instance_group.core-nodes: unexpected EOF
2017/07/27 21:01:06 [TRACE] [walkApply] Exiting eval tree: aws_emr_instance_group.core-nodes
2017/07/27 21:01:06 [DEBUG] dag/walk: upstream errored, not walking "meta.count-boundary (count boundary fixup)"
2017/07/27 21:01:06 [DEBUG] dag/walk: upstream errored, not walking "provider.aws (close)"
2017/07/27 21:01:06 [DEBUG] dag/walk: upstream errored, not walking "root"
2017/07/27 21:01:06 [DEBUG] plugin: ./terraforming/emr/.terraform/plugins/linux_amd64/terraform-provider-aws_v0.1.2_x4: plugin process exited
2017/07/27 21:01:06 [TRACE] Preserving existing state lineage "e613016e-39a8-457a-928a-3332af56661e"
2017/07/27 21:01:06 [TRACE] Preserving existing state lineage "e613016e-39a8-457a-928a-3332af56661e"
2017/07/27 21:01:06 [TRACE] Preserving existing state lineage "e613016e-39a8-457a-928a-3332af56661e"
2017/07/27 21:01:06 [TRACE] Preserving existing state lineage "e613016e-39a8-457a-928a-3332af56661e"
2017/07/27 21:01:06 [DEBUG] plugin: waiting for all plugin processes to complete...
2017/07/27 21:01:06 [WARN] plugin: error closing client during Kill: connection is shut down



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

--- a/aws/resource_aws_emr_cluster.go
+++ b/aws/resource_aws_emr_cluster.go
@@ -360,7 +360,7 @@ func resourceAwsEMRClusterRead(d *schema.ResourceData, meta interface{}) error {
 		d.Set("cluster_state", cluster.Status.State)
 	}
 
-	instanceGroups, err := fetchAllEMRInstanceGroups(meta, d.Id())
+	instanceGroups, err := fetchAllEMRInstanceGroups(emrconn, d.Id())
 	if err == nil {
 		coreGroup := findGroup(instanceGroups, "CORE")
 		if coreGroup != nil {
@@ -414,7 +414,7 @@ func resourceAwsEMRClusterUpdate(d *schema.ResourceData, meta interface{}) error
 	if d.HasChange("core_instance_count") {
 		d.SetPartial("core_instance_count")
 		log.Printf("[DEBUG] Modify EMR cluster")
-		groups, err := fetchAllEMRInstanceGroups(meta, d.Id())
+		groups, err := fetchAllEMRInstanceGroups(conn, d.Id())
 		if err != nil {
 			log.Printf("[DEBUG] Error finding all instance groups: %s", err)
 			return err

--- a/aws/resource_aws_emr_instance_group.go
+++ b/aws/resource_aws_emr_instance_group.go
@@ -155,7 +155,8 @@ func resourceAwsEMRInstanceGroupCreate(d *schema.ResourceData, meta interface{})
 }
 
 func resourceAwsEMRInstanceGroupRead(d *schema.ResourceData, meta interface{}) error {
-	group, err := fetchEMRInstanceGroup(meta, d.Get("cluster_id").(string), d.Id())
+	conn := meta.(*AWSClient).emrconn
+	group, err := fetchEMRInstanceGroup(conn, d.Get("cluster_id").(string), d.Id())
 	if err != nil {
 		switch err {
 		case emrInstanceGroupNotFound:
@@ -186,8 +187,7 @@ func resourceAwsEMRInstanceGroupRead(d *schema.ResourceData, meta interface{}) e
 	return nil
 }
 
-func fetchAllEMRInstanceGroups(meta interface{}, clusterId string) ([]*emr.InstanceGroup, error) {
-	conn := meta.(*AWSClient).emrconn
+func fetchAllEMRInstanceGroups(conn *emr.EMR, clusterId string) ([]*emr.InstanceGroup, error) {
 	req := &emr.ListInstanceGroupsInput{
 		ClusterId: aws.String(clusterId),
 	}
@@ -221,8 +221,8 @@ func fetchAllEMRInstanceGroups(meta interface{}, clusterId string) ([]*emr.Insta
 	return groups, nil
 }
 
-func fetchEMRInstanceGroup(meta interface{}, clusterId, groupId string) (*emr.InstanceGroup, error) {
-	groups, err := fetchAllEMRInstanceGroups(meta, clusterId)
+func fetchEMRInstanceGroup(conn *emr.EMR, clusterId, groupId string) (*emr.InstanceGroup, error) {
+	groups, err := fetchAllEMRInstanceGroups(conn, clusterId)
 	if err != nil {
 		return nil, err
 	}
@@ -280,9 +280,9 @@ func resourceAwsEMRInstanceGroupUpdate(d *schema.ResourceData, meta interface{})
 	return resourceAwsEMRInstanceGroupRead(d, meta)
 }
 
-func instanceGroupStateRefresh(meta interface{}, clusterID, igID string) resource.StateRefreshFunc {
+func instanceGroupStateRefresh(conn *emr.EMR, clusterID, igID string) resource.StateRefreshFunc {
 	return func() (interface{}, string, error) {
-		group, err := fetchEMRInstanceGroup(meta, clusterID, igID)
+		group, err := fetchEMRInstanceGroup(conn, clusterID, igID)
 		if err != nil {
 			return nil, "Not Found", err
 		}

--- a/aws/resource_aws_emr_instance_group_test.go
+++ b/aws/resource_aws_emr_instance_group_test.go
@@ -29,6 +29,27 @@ func TestAccAWSEMRInstanceGroup_basic(t *testing.T) {
 	})
 }
 
+// Confirm we can scale down the instance count. Regression test for https://github.com/terraform-providers/terraform-provider-aws/issues/1264
+func TestAccAWSEMRInstanceGroup_zero_count(t *testing.T) {
+	var ig emr.InstanceGroup
+	rInt := acctest.RandInt()
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSEmrInstanceGroupDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSEmrInstanceGroupConfig(rInt),
+				Check:  testAccCheckAWSEmrInstanceGroupExists("aws_emr_instance_group.task", &ig),
+			},
+			{
+				Config: testAccAWSEmrInstanceGroupConfig_zero_count(rInt),
+				Check:  testAccCheckAWSEmrInstanceGroupExists("aws_emr_instance_group.task", &ig),
+			},
+		},
+	})
+}
+
 func TestAccAWSEMRInstanceGroup_ebsBasic(t *testing.T) {
 	var ig emr.InstanceGroup
 	rInt := acctest.RandInt()
@@ -93,7 +114,8 @@ func testAccCheckAWSEmrInstanceGroupExists(n string, v *emr.InstanceGroup) resou
 			return fmt.Errorf("No task group id set")
 		}
 		meta := testAccProvider.Meta()
-		g, err := fetchEMRInstanceGroup(meta, rs.Primary.Attributes["cluster_id"], rs.Primary.ID)
+		conn := meta.(*AWSClient).emrconn
+		g, err := fetchEMRInstanceGroup(conn, rs.Primary.Attributes["cluster_id"], rs.Primary.ID)
 		if err != nil {
 			return fmt.Errorf("EMR error: %v", err)
 		}
@@ -176,6 +198,10 @@ resource "aws_security_group" "allow_all" {
 resource "aws_vpc" "main" {
   cidr_block           = "168.31.0.0/16"
   enable_dns_hostnames = true
+
+	tags {
+		Name = "tf_acc_emr_tests"
+	}
 }
 
 resource "aws_subnet" "main" {
@@ -375,7 +401,17 @@ func testAccAWSEmrInstanceGroupConfig(r int) string {
 	resource "aws_emr_instance_group" "task" {
     cluster_id     = "${aws_emr_cluster.tf-test-cluster.id}"
     instance_count = 1
-    instance_type  = "m3.xlarge"
+    instance_type  = "m1.small"
+  }
+	`, r, r, r, r, r, r)
+}
+
+func testAccAWSEmrInstanceGroupConfig_zero_count(r int) string {
+	return fmt.Sprintf(testAccAWSEmrInstanceGroupBase+`
+	resource "aws_emr_instance_group" "task" {
+    cluster_id     = "${aws_emr_cluster.tf-test-cluster.id}"
+    instance_count = 0
+    instance_type  = "m1.small"
   }
 	`, r, r, r, r, r, r)
 }

